src/adc.rs

```diff
@@ -27,358 +27,195 @@ static mut SPI_START: [u16; 1] = [0x00];
-                // actually modified. It technically only needs to be immutably borrowed, but the
-                // current HAL API only supports mutable borrows.
-                unsafe { &mut SPI_START },
-                None,
-                trigger_config,
-            );
-
-        // The data stream constantly reads from the SPI RX FIFO into a RAM buffer. The peripheral
-        // stalls reads of the SPI RX FIFO until data is available, so the DMA transfer completes
-        // after the requested number of samples have been collected. Note that only ADC1's data
-        // stream is used to trigger a transfer completion interrupt.
-        let data_config = DmaConfig::default()
-            .memory_increment(true)
-            .priority(Priority::VeryHigh);
-
-        // A SPI peripheral error interrupt is used to determine if the RX FIFO overflows. This
-        // indicates that samples were dropped due to excessive processing time in the main
-        // application (e.g. a second DMA transfer completes before the first was done with
-        // processing). This is used as a flow control indicator to guarantee that no ADC samples
-        // are lost.
-        let mut spi = spi.disable();
-        spi.listen(hal::spi::Event::Error);
-
-        // The data transfer is always a transfer of data from the peripheral to a RAM buffer.
-        let mut data_transfer: Transfer<_, _, PeripheralToMemory, _> =
-            Transfer::init(
-                data_stream,
-                spi,
-                // Note(unsafe): The ADC0_BUF0 is "owned" by this peripheral. It shall not be used
-                // anywhere else in the module.
-                unsafe { &mut ADC0_BUF0 },
-                None,
-                data_config,
-            );
-
-        data_transfer.start(|spi| {
-            // Allow the SPI FIFOs to operate using only DMA data channels.
-            spi.enable_dma_rx();
-            spi.enable_dma_tx();
-
-            // Enable SPI and start it in infinite transaction mode.
-            spi.inner().cr1.modify(|_, w| w.spe().set_bit());
-            spi.inner().cr1.modify(|_, w| w.cstart().started());
-        });
-
-        trigger_transfer.start(|_| {});
-
-        Self {
-            // Note(unsafe): The ADC0_BUF1 is "owned" by this peripheral. It shall not be used
-            // anywhere else in the module.
-            next_buffer: unsafe { Some(&mut ADC0_BUF1) },
-            transfer: data_transfer,
-            _trigger_transfer: trigger_transfer,
+static mut ADC_BUF: [[[u16; SAMPLE_BUFFER_SIZE]; 2]; 2] =
+    [[[0; SAMPLE_BUFFER_SIZE]; 2]; 2];
+
+macro_rules! adc_input {
+    ($name:ident, $index:literal, $trigger_stream:ident, $data_stream:ident,
+     $spi:ident, $trigger_channel:ident, $dma_req:ident) => {
+        /// $spi is used as a ZST (zero-sized type) for indicating a DMA transfer into the SPI TX FIFO
```

Comment:
Nit: SPI is no longer a ZST, since it contains the (unused) timer channel for safety guarantees. (This was not introduced here, but we can fix it while updating this).